### ISSUE TITLE
UNIX sockets in abstract namespace not supported

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4421]
|<img alt="cgay's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| cgay reported|
|-|-|
|Trac ID|trac#4421|
|Type|defect|
|Created|2010-05-04 17:49:47Z|
|Branch|https://github.com/twisted/twisted/tree/abstract-unix-4421-3|

Twisted 9.0 and 10.0

Given this diff...

```
Index: doc/core/examples/simpleserv.py
===================================================================
--- doc/core/examples/simpleserv.py	(revision 28890)
+++ doc/core/examples/simpleserv.py	(working copy)
@@ -18,7 +18,7 @@
     """This runs the protocol on port 8000"""
     factory = protocol.ServerFactory()
     factory.protocol = Echo
-    reactor.listenTCP(8000,factory)
+    reactor.listenUNIX('\0xxx', factory)
     reactor.run()
 
 # this only runs if the module was *not* imported

```

I get this error...

```
$ python simpleserv.py 
Traceback (most recent call last):
  File "simpleserv.py", line 26, in <module>
    main()
  File "simpleserv.py", line 21, in main
    reactor.listenUNIX('\0xxx', factory)
  File ".../python2.6/site-packages/twisted/internet/posixbase.py", line 314, in listenUNIX
    p.startListening()
  File ".../python2.6/site-packages/twisted/internet/unix.py", line 95, in startListening
    os.chmod(self.port, self.mode)
TypeError: chmod() argument 1 must be encoded string without NULL bytes, not str
```

Attempting to change the mode of an abstract UNIX socket is probably not so good.  Some discussion in #twisted yesterday (2010.may.03).

Note that it probably also blows up on os.unlink(self.port).

-Carl

<details><summary>Searchable metadata</summary>

```
trac-id__4421 4421
type__defect defect
reporter__cgay cgay
priority__normal normal
milestone__ 
branch__branches_abstract_unix_4421_3 branches/abstract-unix-4421-3
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy_ easy,
time__1272995387000000 1272995387000000
changetime__1275928948000000 1275928948000000
version__None None
owner__ 
cc__cgay
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="cgay's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|cgay commented|
|-|-|

Probably at the same time it would be good to fix the "(Port None Closed)" log messages as well:

```
2010-05-05 13:04:24-0400 [-] Log opened.
2010-05-05 13:04:24-0400 [-] twistd 10.0.0 (/ita/installs/smpython-2.6/bin/python 2.6.4) starting up.
2010-05-05 13:04:24-0400 [-] reactor class: twisted.internet.selectreactor.SelectReactor.
2010-05-05 13:04:24-0400 [-] twisted.web.server.Site starting on 5555
2010-05-05 13:04:24-0400 [-] Starting factory <twisted.web.server.Site instance at 0x2aaab24a1fc8>
2010-05-05 13:04:24-0400 [-] twisted.web.server.Site starting on "'\\x00unix-socket-5555'"
2010-05-05 13:04:24-0400 [-] twisted.web.server.Site starting on 7555
2010-05-05 13:04:24-0400 [-] twisted.web.server.Site starting on "'\\x00unix-socket-7555'"
2010-05-05 13:04:24-0400 [-] startService
2010-05-05 13:08:34-0400 [-] Received SIGINT, shutting down.
2010-05-05 13:08:34-0400 [-] (Port None Closed)
2010-05-05 13:08:34-0400 [-] (Port 7555 Closed)
2010-05-05 13:08:34-0400 [-] (Port None Closed)
2010-05-05 13:08:34-0400 [-] (Port 5555 Closed)
2010-05-05 13:08:34-0400 [-] Stopping factory <twisted.web.server.Site instance at 0x2aaab24a1fc8>
2010-05-05 13:08:34-0400 [-] Main loop terminated.
2010-05-05 13:08:34-0400 [-] Server Shut Down.

```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [29138]) Branching to 'abstract-unix-4421'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [29151]) Branching to 'abstract-unix-4421-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

Filed [#4477](https://github.com/twisted/twisted/issues/4477) for the logging issue.  The rest ought to be fixed in the branch.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/abstract-unix-4421-2).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [29154]) Fix mode argument, which now actually triggers a test failure after a blanket try/except has been removed

refs [#4421](https://github.com/twisted/twisted/issues/4421)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @exarkun|
|-|-|

Waiting for [#4078](https://github.com/twisted/twisted/issues/4078) on this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [29164]) Branching to 'abstract-unix-4421-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun removed owner|
|-|-|

Okay, here we go again.  [Latest build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/abstract-unix-4421-3).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @exarkun|
|-|-|

Garr.  Jumped the gun, again, apparently.  Some test failures on buildbot, looking at them now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun removed owner|
|-|-|

Okay.  I fixed some Python 2.4 issues and now I think things are happy.

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TimAllen set owner to TimAllen|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|TimAllen set owner to @exarkun|
|-|-|

Looks pretty good to me.

The only problem I have is that if abstract-namespace sockets are a Linux-specific extension, the particular details might not be familiar to later maintainers, and all the "`self.port[:1] != '\0'`" tests will likely seem pretty crazy.

I suggest using a helper function:
```
def _socketInFilesystemNamespace(socketpath):
    """
    Test if the given socket path is in filesystem namespace.

    While most PF_UNIX sockets are entries in the filesystem, Linux 2.2 and above
    support PF_UNIX sockets in an "abstract namespace" that does not correspond to 
    any path. This method returns C{True} if the given socket path is stored in the
    filesystem, and C{False} otherwise.
    """
    return socketpath[:1] != "\0"
```
...and changing the tests to use it:
```
if _socketInFilesystemNamespace(self.port):
    # Make the socket readable and writable to the world.
    os.chmod(self.port, self.mode)
```

That way, anyone who's curious about the possibility of sockets existing outside the file-system name-space can go read the docstring and be enlightened.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [29187]) Factor nul check into a documented helper

refs [#4421](https://github.com/twisted/twisted/issues/4421)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun removed owner|
|-|-|

Thanks for the review.  That's an excellent suggestion to improve the readability of this code, so I've adopted it almost verbatim.

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TimAllen set owner to TimAllen|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|TimAllen set owner to @exarkun|
|-|-|

+1 for merging.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set status to `closed`|
|-|-|

(In [29240]) Merge abstract-unix-4421-3

Author: exarkun
Reviewer: TimAllen
Fixes: [#4421](https://github.com/twisted/twisted/issues/4421)

Support abstract namespace unix sockets on Linux by avoiding applying filesystem
operations to them.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
